pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -0,0 +1,743 @@
+		if n == nodeName {
+			return true
+		}
+	}
+	return false
+}
+
+// Reserve reserves claims for the pod.
+func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
+	state, err := getStateData(cs)
+	if err != nil {
+		return statusError(klog.FromContext(ctx), err)
+	}
+	if len(state.claims) == 0 {
+		return nil
+	}
+
+	pending := 0
+	infos := 0
+	logger := klog.FromContext(ctx)
+	podScheduling, err := state.initializePodScheduling(ctx, pod, pl.podSchedulingLister)
+	if err != nil {
+		return statusError(logger, err)
+	}
+	for index, claim := range state.claims {
+		if claim.Status.Allocation != nil {
+			// Allocated, but perhaps not reserved yet.
+			if resourceclaim.IsReservedForPod(pod, claim) {
+				logger.V(5).Info("is reserved", "pod", klog.KObj(pod), "node", klog.ObjectRef{Name: nodeName}, "resourceclaim", klog.KObj(claim))
+				continue
+			}
+			claim := claim.DeepCopy()
+			claim.Status.ReservedFor = append(claim.Status.ReservedFor,
+				v1.ResourceClaimUserReference{
+					Resource: "pods",
+					Name:     pod.Name,
+					UID:      pod.UID,
+				})
+			logger.V(5).Info("reserve", "pod", klog.KObj(pod), "node", klog.ObjectRef{Name: nodeName}, "resourceclaim", klog.KObj(claim))
+			_, err := pl.clientset.CoreV1().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+			// TODO: metric for update errors.
+			if err != nil {
+				return statusError(logger, err)
+			}
+			// If we get here, we know that reserving the claim for
+			// the pod worked and we can proceed with scheduling
+			// it.
+		} else {
+			// Must be delayed allocation.
+			pending++
+
+			// Did the driver provide information that steered node
+			// selection towards a node that it can support?
+			if statusForClaim(podScheduling, pod.Spec.ResourceClaims[index].Name) != nil {
+				infos++
+			}
+		}
+	}
+
+	if pending == 0 {
+		// Nothing left to do.
+		return nil
+	}
+
+	podSchedulingDirty := state.podSchedulingDirty
+	if len(podScheduling.Spec.PotentialNodes) == 0 {
+		// PreScore was not called, probably because there was
+		// only one candidate. We need to ask whether that
+		// node is suitable, otherwise the scheduler will pick
+		// it forever even when it cannot satisfy the claim.
+		podScheduling = podScheduling.DeepCopy()
+		podScheduling.Spec.PotentialNodes = []string{nodeName}
+		logger.V(5).Info("asking for information about single potential node", "pod", klog.KObj(pod), "node", klog.ObjectRef{Name: nodeName})
+		podSchedulingDirty = true
+	}
+
+	// When there is only one pending resource, we can go ahead with
+	// requesting allocation even when we don't have the information from
+	// the driver yet. Otherwise we wait for information before blindly
+	// making a decision that might have to be reversed later.
+	if pending == 1 || infos == pending {
+		podScheduling = podScheduling.DeepCopy()
+		podScheduling.Spec.SelectedNode = nodeName
```

Comment:
The scheduler is not guaranteed to select the same node every cycle.

What happens if a driver already started allocating and we change this value? Could we end up scheduling the pod in a different node where the resource is not allocated?

Reply:
No. The scheduler will check the actual allocation once that is complete before scheduling the pod.

But selecting a different node while allocation is on-going isn't ideal because it could lead to the situation where some claims are allocated for one node and some others for another. I have added a TODO to investigate whether this can be avoided.

I'm not sure how common it will be: if pod scheduling stops because it had to trigger delayed allocation, won't the pod remain on hold until the ResourceClaim change triggers another scheduling attempt, or at least long enough for allocation to complete? I see that it's not guaranteed, but perhaps it is common enough to be not a big issue?

There is the deallocation feature that will deal with the situation when it goes wrong, too.

Reply:
You can control retries based on events.

But the question remains: what if the claim is allocated, causing the scheduler to run the scheduling cycle again, but it decides to pick a different node? Could that happen? Sure, it might not actually schedule the pod there, but would it try to deallocate to assign to the new selected node?

Reply:
> but it decides to pick a different node? Could that happen?

No, it can't. At that time the Allocation is set and, if the resource is limited to the originally selected node, will have the AvailableOnNodes selector set, so the scheduler is force to pick that node again.

Only if the result of the scheduling attempt is that the node doesn't fit the pod anymore and some other node would, then deallocation kicks in.